source/simulator/core.cc

```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+                            {
+                              parameters.linear_stokes_solver_tolerance = std::min(parameters.minimum_linear_stokes_solver_tolerance,
+                                                                                   std::fabs(newton_residual-stokes_residual)/(newton_residual_old));
+                            }
+                          else
+                            {
+                              parameters.linear_stokes_solver_tolerance = std::min(parameters.minimum_linear_stokes_solver_tolerance,
+                                                                                   std::max(std::fabs(newton_residual-stokes_residual)/newton_residual_old,
+                                                                                            std::pow(parameters.linear_stokes_solver_tolerance,powerterm)));
+                            }
+                        }
+                      else
+                        {
+                          if (0.9*parameters.linear_stokes_solver_tolerance * parameters.linear_stokes_solver_tolerance <= 0.1)
+                            {
+                              parameters.linear_stokes_solver_tolerance =  std::min(parameters.minimum_linear_stokes_solver_tolerance,
+                                                                                    0.9 * std::fabs(newton_residual * newton_residual) /
+                                                                                    (newton_residual_old * newton_residual_old));
+                            }
+                          else
+                            {
+                              parameters.linear_stokes_solver_tolerance = std::min(parameters.minimum_linear_stokes_solver_tolerance,
+                                                                                   std::max(0.9 * std::fabs(newton_residual*newton_residual) / (newton_residual_old*newton_residual_old),
+                                                                                            0.9*parameters.linear_stokes_solver_tolerance*parameters.linear_stokes_solver_tolerance));
+                            }
+                        }
+                      pcout << "   The linear solver tolerance is set to " << parameters.linear_stokes_solver_tolerance << std::endl;
+                    }
+                }
+
+              build_stokes_preconditioner();
+
+              stokes_residual = solve_stokes();
+
+              newton_residual_velo = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+              newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+              newton_residual = std::sqrt(newton_residual_velo * newton_residual_velo + newton_residual_pres * newton_residual_pres);
+
+              // We may need to do a line search if the solution doesn't decrease the norm of the rhs. TODO: more documentation.
```

Comment:
see TODO. In particular: Why do you need a backup copy of the current_linearization_point for the line search?